### PR TITLE
[incubator/kafka] Added support for Prometheus exporter, updated JMX stats

### PR DESCRIPTION
**Pull request by bradenwright**
**What this PR does / why we need it**:
Adding support for Prometheus Operator, and cleaning up JMX stats to be more Prometheus style.  Added some variables, updated jmx configmap to have more option for how to override content.  Allows JMX whitelist objects to be set.

### CODE REVIEW

**Comment by bradenwright** [comment]
Alright so I got together a PR to support Prometheus Operator for stats.  I made some other updates as well.  Mainly cleaning up JMX stats and adding some vars.  In the future it would be nice to whitelist and clean up more jmx stats yet, but that would be a separate PR.  I also plan on releasing in some fashion a copy of the Grafana Dashboard we are using.  

If someone could verify that it works without the Prometheus Operator, I tested as much of the code as I could but wasn't able to test that.  @nrmitchi I know you are using it that way, would be great if you could assist me.

@benjigoldberg @nrmitchi @t-d-d @ebabani @aruneli  please comment, yell, swear, thumbs up, etc.  I more than happy to discuss any changes.

File: incubator/kafka/README.md
Comment (benjigoldberg) [inline incubator/kafka/README.md]: formatting in markdown breaks here

File: incubator/kafka/values.yaml
Comment (benjigoldberg) [inline incubator/kafka/values.yaml]: can you tell us a little more about this change and why there are two ports?

Comment (bradenwright) [reply incubator/kafka/values.yaml]: Sure, I renamed to try and make the distinction b/t the 2 ports a little more clear.  There is a top level `jmxPort` var now which is the actually jmx port that exposes metrics in jmx format.  The JMX exporter for prometheus transforms those stats into a prometheus format and exposes them for scraping at `prometheus.jmx.port`

I had a couple question on it, so I thought renaming `metrics` to `prometheus` made things more clear.  If someone has a better suggestion let me know, but I think its clearer since metrics could be jmx or prometheus style.

File: incubator/kafka/values.yaml
Comment (ebabani) [inline incubator/kafka/values.yaml]: Indicate it's for prometheus operator

File: incubator/kafka/values.yaml
Comment (ebabani) [inline incubator/kafka/values.yaml]: Default to false

**Comment by ebabani** [comment]
Change looks good, but I'm a bit hesitant adding prometheus operator specific resources since it's not part of core kubernetes. I also didn't see any other charts with explicit support for prometheus operator.

Not saying don't do it, but I'd default to having prometheus operator disabled if metrics are on.

File: incubator/kafka/values.yaml
Comment (bradenwright) [reply incubator/kafka/values.yaml]: Prometheus Operator is the standard that's recommended by CoreOS, I feel it should be defaulted to yes IMHO

**Comment by bradenwright** [comment]
I feel like if you are starting out fresh you should / would most likely use the Prometheus Operator over Prometheus.  So I defaulted to yes.  That's my opinion, but its a democracy so if I get overruled, not my hill to die on.

File: incubator/kafka/values.yaml
Comment (nrmitchi) [reply incubator/kafka/values.yaml]: I'm also not a fan of this defaulting as `true`; even though it's recommended by CoreOS, IIRC CoreOS isn't the owner/maintainer of any of the components of this (Kubernetes, Prometheus, or Helm)

It's also not the default way that many Helm users would get prometheus (the default `prometheus` chart doesn't use the operator), so does not seem like it should be the default in this sense.

File: incubator/kafka/values.yaml
Comment (nrmitchi) [inline incubator/kafka/values.yaml]: This doesn't seem like something that is mean to be under the configMap key?

I think I'm not sure why all of this section seems to be under a `configMap` key at all (rather than just directly under `jmx`.

As well, alot of this seems like it is effectively the same as what was previous under the `metrics` key, just moved?

File: incubator/kafka/templates/jmx-configmap.yaml
Comment (nrmitchi) [inline incubator/kafka/templates/jmx-configmap.yaml:43]: On the previous PR I talked about why I'm not really a fan of rewriting metric names (https://github.com/kubernetes/charts/pull/4138#discussion_r180472559)

Granted, I am not really familiar with jmx configuration, but will using all of these rules as defaults potentially prevent users from being able to effectively use existing JMX/monitoring tutorials? (metrics not existing as they 'should', etc?)

I think providing education about how to configure this appropriately is good, but I'm not sure that complex rules should be part of the default. Perhaps this configuration information, or a link to it, would be best served as part of the NOTES.txt template?

**Review by nrmitchi — COMMENTED** [review]
Overall I feel that if more config is added to support prometheus-operator specifically, it would be best living under a new key in the configuration; likely under `metrics.prometheusOperator`, and attempt to limit prometheus-operator specific keys leaking into other areas (such as these interval keys: https://github.com/kubernetes/charts/pull/5120/files#diff-8b7f371fb6e4a969043490ec93e18495R214)

As well, some of the other changes in this PR (such as https://github.com/kubernetes/charts/pull/5120/files#diff-8b7f371fb6e4a969043490ec93e18495R217) are good, but seem unrelated, and may be better suited in their own PR.

File: incubator/kafka/values.yaml
Comment (aruneli) [reply incubator/kafka/values.yaml]: IMHO, we can leave it true. I and many others I know of, by default use operator not the default prometheus chart. My 2 cents.

File: incubator/kafka/templates/jmx-configmap.yaml
Comment (bradenwright) [reply incubator/kafka/templates/jmx-configmap.yaml:43]: This is something I feel very strongly about.  The way metrics are exposed doesn't work well in Prometheus.  If you spend a few mins looking at the stats they are super ugly, doing things like moving topics to labels 💯 is something that should be done.

If you want to change some of the default jmx configuration I'm open, but it should ship with a standard that starts users off in a good place.

File: incubator/kafka/values.yaml
Comment (bradenwright) [reply incubator/kafka/values.yaml]: It's intentional, allowing people to disable the configmap and provide their own can be very useful.  I'll see if I can update the comments in a way to make it more clear.

File: incubator/kafka/templates/jmx-configmap.yaml
Comment (bradenwright) [reply incubator/kafka/templates/jmx-configmap.yaml:43]: Especially since I plan to publish a Grafana dashboard to https://grafana.com/dashboards and that's based on the formatting of the stats.  I by no means have all the answers to what should be monitored and best dashboards, but I want to provide a starting point and let the community help improve them (how to clean up stats too)

Comment (nrmitchi) [reply incubator/kafka/templates/jmx-configmap.yaml:43]: I agree with your points, I'm just concerned that would then not be clear how to use the metrics without digging into the chart, or diving in to find the configuration applied. 

This seems like a good use of the `NOTES.txt` section; how about we add a note that default JMX configuration has been added, along with instructions of how to view it? (ie, a copy/paste-able `kubectl` command to view the configmap)

Comment (bradenwright) [reply incubator/kafka/templates/jmx-configmap.yaml:43]: I like that suggestion, anything we can do to be clear without making people dig.  

I'm actually trying to get jconsole (comes with java, gui I found with a quick google) connecting, b/c when I was creating the config file I was port forward with kubectl and  looking at both raw metrics connection to `.Values.jmx.port` via jconsole and looking at scraped metrics `Values.prometheus.jmx.port` in a web browser (port forwarding to access them).  So I'm trying to get that going again but having a little trouble, if I get it working I can try to put the port-forward commands for those in NOTES.txt too.

File: incubator/kafka/values.yaml
Comment (ebabani) [reply incubator/kafka/values.yaml]: Doesn't feel right to me to have a chart rely on a third party product and Custom Resources that don't come standard with Kubernetes.

I'm interested in seeing if e2e tests will handle it though.

**Comment by t-d-d** [comment]
Maybe do this as a separate chart?

**Comment by bradenwright** [comment]
Unless I'm not following.   I don't see how this can be broken into a separate chart, the jmx config file is a side car container in the same pod as kafka to monitor jmx.

**Comment by aruneli** [comment]
I manually imported the changes from this PR and tested end-end and it worked fine. I am happy to test it more if there is any specific requirement.

File: incubator/kafka/templates/jmx-configmap.yaml
Comment (benjigoldberg) [reply incubator/kafka/templates/jmx-configmap.yaml:43]: Nice suggestion on the `NOTES.txt` front, I think that will make things nice and clear for the end user

File: incubator/kafka/values.yaml
Comment (benjigoldberg) [inline incubator/kafka/values.yaml]: Can we change the wording here a bit -- it took me a read or two to parse what this meant. I think something along the lines of:

```
## If you would like to supply your own ConfigMap for JMX metrics, supply the name of that 
## ConfigMap as an `overrideName` here.
```

File: incubator/kafka/values.yaml
Comment (benjigoldberg) [inline incubator/kafka/values.yaml]: If somebody wanted to expose all the metrics how would they do that? Perhaps document that in the comment.

File: incubator/kafka/values.yaml
Comment (benjigoldberg) [reply incubator/kafka/values.yaml]: @viglesiasce I'm curious if you have a perspective on using Prometheus Operator as a default pattern in one of our charts? I am personally a fan of the operator pattern, but I wonder if perhaps its a little inconsistent to allow this as a default given that we don't offer a chart for Prometheus Operator in this repository?

I'm leaning towards defaulting this to false personally.

**Review by benjigoldberg — APPROVED** [review]

**Comment by ebabani** [comment]
Looks like there's some hardcoded namespaces.

https://github.com/kubernetes/charts/pull/5120/files#diff-6bd910f600a63e4f5b84e39c42c1bb23R6

**Comment by bradenwright** [comment]
@ebabani good point, it needs to be in same namespace as prometheus. But it should be configurable, changing now.

**Comment by benjigoldberg** [comment]
@bradenwright @ebabani @nrmitchi @t-d-d my thought is this -- lets default the operator to false. We can merge this PR see if there is a community call for changing the default to true.

What I would like to do though, is add a section in the README for details on monitoring. I think given the extent of the features on offer here, we should make sure to document how these features work. I can imagine users who werent part of these PR discussions over the past month or so having a hard time understanding what to expect without diving into the charts.

File: incubator/kafka/README.md
Comment (benjigoldberg) [inline incubator/kafka/README.md]: Standard Prometheus is the default monitoring option for this chart. This chart also supports the CoreOS Prometheus Operator, which can provide additional functionality like automatically updating Prometheus and Alert Manager configuration. If you are interested in installing the Prometheus Operator please see the [CoreOS repository for more information](https://github.com/coreos/prometheus-operator/tree/master/helm) or read through the [CoreOS blog post introducing the Prometheus Operator](https://coreos.com/blog/the-prometheus-operator.html)

File: incubator/kafka/README.md
Comment (benjigoldberg) [inline incubator/kafka/README.md]: The majority of Kafka statistics are provided via JMX and are exposed via the [Prometheus JMX Exporter](https://github.com/prometheus/jmx_exporter)

File: incubator/kafka/README.md
Comment (benjigoldberg) [inline incubator/kafka/README.md]: The JMX Exporter is a general purpose prometheus provider which is intended for use with any Java application. Because of this, it produces a number of statistics which may not be of interest. To help in reducing these statistics to their relevant components we have created a curated white list for the JMX exporter. This white list may be modified or removed via the values configuration.

File: incubator/kafka/README.md
Comment (benjigoldberg) [inline incubator/kafka/README.md]: To accommodate compatibility with the Prometheus collector, this chart performs transformations on top of the raw JMX metrics. For example, broker names and topics names are incorporated into the metric name instead of becoming a label. If you are curious to learn more about any default transformations to the chart metrics, please have [reference the configmap template](https://github.com/kubernetes/charts/blob/master/incubator/kafka/templates/jmx-configmap.yaml).

File: incubator/kafka/README.md
Comment (benjigoldberg) [inline incubator/kafka/README.md]: The [Kafka Exporter](https://github.com/danielqsj/kafka_exporter) is a complimentary metrics exporter to the JMX Exporter. The Kafka Exporter provides additional statistics on Kafka Consumer Groups.

**Review by benjigoldberg — CHANGES_REQUESTED** [review]

File: incubator/kafka/values.yaml
Comment (nrmitchi) [inline incubator/kafka/values.yaml]: Small nit: Do these instructions exist in a separate place already? If so, would linking to them be better? I would worry that duplicating instructions for installing an operator here could easily get out of date

**Pull request merged**
